File: src/main/java/bot/commands/vc/play/MusicPlayer.java
package bot.commands.vc.play;

import bot.commands.Command;
import bot.utils.EventUtils;
import com.sedmelluq.discord.lavaplayer.player.AudioLoadResultHandler;
import com.sedmelluq.discord.lavaplayer.player.AudioPlayer;
import com.sedmelluq.discord.lavaplayer.player.AudioPlayerManager;
import com.sedmelluq.discord.lavaplayer.player.DefaultAudioPlayerManager;
import com.sedmelluq.discord.lavaplayer.source.AudioSourceManagers;
import com.sedmelluq.discord.lavaplayer.tools.FriendlyException;
import com.sedmelluq.discord.lavaplayer.track.AudioPlaylist;
import com.sedmelluq.discord.lavaplayer.track.AudioTrack;
import com.sedmelluq.discord.lavaplayer.track.playback.NonAllocatingAudioFrameBuffer;
import net.dv8tion.jda.api.Permission;
import net.dv8tion.jda.api.entities.GuildVoiceState;
import net.dv8tion.jda.api.entities.Member;
import net.dv8tion.jda.api.entities.VoiceChannel;
import net.dv8tion.jda.api.events.Event;
import net.dv8tion.jda.api.events.message.MessageReceivedEvent;
import net.dv8tion.jda.api.managers.AudioManager;
import net.dv8tion.jda.api.utils.WidgetUtil;

import javax.sound.midi.Track;
import java.util.Objects;

public class MusicPlayer extends Command {
    public MusicPlayer() {
        super(false, "play", "Plays music etc..", "p");
    }

    @Override
    public void handle(MessageReceivedEvent event) {
        if (event.getAuthor().isBot()) return;

        GuildVoiceState state = Objects.requireNonNull(event.getMember()).getVoiceState();
        if (state == null) return;

        VoiceChannel vc = state.getChannel();
        if (vc == null) return;

        if (!event.getGuild().getSelfMember().hasPermission(vc, Permission.VOICE_CONNECT)) return;

        AudioManager audioManager = event.getGuild().getAudioManager();

        if (audioManager.isConnected()) {
            EventUtils.sendMessageToChannel(event, "Bot is already connected to a voice channel!");
            return;
        }

        audioManager.openAudioConnection(vc);
        EventUtils.sendMessageToChannel(event, "Connected to channel!");

        AudioPlayerManager playerManager = new DefaultAudioPlayerManager();
        AudioSourceManagers.registerRemoteSources(playerManager);

        AudioPlayer player = playerManager.createPlayer();

        TrackScheduler trackScheduler = new TrackScheduler(player);

        playerManager.loadItem("player", trackScheduler);

        
    }
}

File: src/main/java/bot/commands/vc/play/MusicPlayer.kt
package bot.commands.vc.play

import bot.Bot
import bot.commands.Command
import bot.utils.EventUtils
import com.sedmelluq.discord.lavaplayer.player.AudioLoadResultHandler
import com.sedmelluq.discord.lavaplayer.player.AudioPlayerManager
import com.sedmelluq.discord.lavaplayer.player.DefaultAudioPlayerManager
import com.sedmelluq.discord.lavaplayer.source.AudioSourceManagers
import com.sedmelluq.discord.lavaplayer.tools.FriendlyException
import com.sedmelluq.discord.lavaplayer.track.AudioPlaylist
import com.sedmelluq.discord.lavaplayer.track.AudioTrack
import net.dv8tion.jda.api.Permission
import net.dv8tion.jda.api.events.message.MessageReceivedEvent
import java.util.*

class MusicPlayer(val bot: Bot) : Command(bot, false, "play", "Plays music etc..", "p") {
    override fun handle(event: MessageReceivedEvent) {
        println("Handle Music Play")
        if (event.author.isBot) return

        println("Before Voice State")
        val state = Objects.requireNonNull(event.member)!!.voiceState ?: return

        println("Before Channel Check")
        val vc = state.channel ?: return

        val link = bot.lavalink.getLink(event.guild)
        link.connect(vc)

        val track = EventUtils.spitRawContent(event)[1]

        bot.playerManager.loadItem(track, object : AudioLoadResultHandler {
            override fun loadFailed(e: FriendlyException) {
                event.channel.sendMessage(e.message ?: e.toString()).queue()
            }

            override fun trackLoaded(track: AudioTrack) {
                event.channel.sendMessage("Playing `${track.info.title}`").queue()
                link.player.playTrack(track)
            }

            override fun noMatches() {
                event.channel.sendMessage("No matches").queue()
            }

            override fun playlistLoaded(playlist: AudioPlaylist) {
                val loaded = playlist.selectedTrack ?: playlist.tracks.firstOrNull()
                if (loaded == null) {
                    event.channel.sendMessage("Empty playlist").queue()
                    return
                }
                event.channel.sendMessage("Playing `${loaded.info.title}` from list `${playlist.name}`").queue()
                link.player.playTrack(loaded)
            }
        })
    }
}